base-stats: pystats refactoring and CSV stats implementation

jonathanhaley opened the pull request:
This series of patches aims simplify the development of new statistics output formats by providing processing methods for each statistic type. 

 It also includes a new visitor object to output stats in a CSV format.

jonathanhaley committed:
base-stats, python: remove setattr from group stats

To add clarity to hierarchical stats we dont allow arbitrary
variable assignment. Instead we add statistics to a dictionary
called values. This has the added benefit that we can also
descriminate between variables speicific to the group and
variables that would be related to sub-stats.

Change-Id: I4b7a30bfe63da336e9c83d84945ad22f9a65c7b5
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

jonathanhaley committed:
base-stats, python: move stat processing out of serializable stat

Python Stats processing is inherent to the type of stat-processing visitor
rather than the stat itself. Therefore we move stat-porcessing out of the
serializable stats class and into each visitor to determine how each
statistic type should be dealt with. This simplifies the process for
defining new output formats as we now just describe the required
processing for each case type individually:

Scalar, Distribution and SparseHist for specfic values.
Vector and Vector 2d to hold sets of result values.
SimObjectGroups as well as the overall SimStat object in order to
manage the overall heirarchical structure of statistics.

Finally we define a dump method to write the processed statistics to
a file.

Change-Id: I4210090fd860908f14702d0a049068a8c8e48e34
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

jonathanhaley committed:
base-stats: add name attribute to Group

Most stats Groups have a name attribute so we add it to the base stats group.

Change-Id: I4cdce19950137f6d79d7cff4836362b7862c47f7
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

jonathanhaley committed:
base-stats, python: Add CSV stats format

Added a CsvOutputVistor to allow stats to be output in a
CSV format with one row per stats dump.
CSV stats can be added by calling "--stats-file=csv://filename.csv"

Change-Id: I47af8d94b8a2788eb0a29c6c8d1c215ff22fc3f9
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

jonathanhaley committed:
base-stats: remove Serializable Stat class

The change "move stat processing out of serializable stat" makes  the
serializable stat class redundant. We can therefore remove it.

Change-Id: I3d36c1f1672b17a40081e5a7f820574d55da7f25
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

powerjg commented:
Thanks for the contribution! This is a good feature to add. We'll get back to this after the release, if that's OK with you. If we don't respond to this in a couple of weeks, please ping again 🙂

jonathanhaley committed:
base-stats, python: fixup group accesses with getattr

Change-Id: I71a14b61af7f0f3532ea0bcc96c4b76f4acc2c79

jonathanhaley committed:
tests: fixup tests for stat changes

Modified some tests which were accessing stats in json format without
using the json visitor object. Also modified the simobjectVectorTest
as modifications to the structure of simobjects, some stats are now
accessed differently  better aligning to the output given when
outputting as a json object. E.g. in the test when accessing
stat_testers1's description by calling:

simstat.stat_testers.value[1].placeholder.description
which aligns to the json structure:

    "stat_testers": {
            "type": "SimObjectVector",
            "time_conversion": null,
            "value": [
                {
                    ...
                {
                    "type": "SimObject",
                    "time_conversion": null,
                    "name": "stat_testers1",
                    "placeholder": {
                        "value": 22.0,
                        "type": "Scalar",
                        "description": "Index 2 desc.",
                        "unit": "Count",
                        "datatype": "f64"
                    }
                }]

Change-Id: I638df04483a9aa0b4557859b63eda888a80a8a82

BobbyRBruce commented:
Hey @jonathanhaley: It seems like this change is causing test failures. To start, could you rebase on top of the current gem5 develop branch (or merge the gem5 develop branch into these changes)? It appears there are conflicts which will also need to be resolved when doing this.

powerjg commented on src/python/m5/stats/gem5stats.py:
```diff
@@ -59,7 +65,210 @@
+            Group,
+        ]:
+            return True
+        print(type(element))
```
Is this a debug print?

jonathanhaley replied:
Whoops, yes it was a print while I was fixing some issues. I have removed it now.

powerjg commented on src/python/m5/stats/gem5stats.py:
```diff
@@ -59,7 +65,210 @@
+class CsvOutputVisitor(Visitor):
+    """
+    This is a helper vistor class used to include a CSV output via the stats
+    API (``src/python/m5/stats/__init__.py``).
```
Can you describe what the output format looks like here (or somewhere else)?

jonathanhaley replied:
Yes, have done. Hopefully the following is clear enough.

powerjg reviewed (commented):
lgtm

jonathanhaley committed:
base-stats, python: remove debug print from gem5stats function

jonathanhaley committed:
base-stats, python: Add CSV stats format description to visitor

jonathanhaley committed:
base-stats, python: Store multiple dumps within csv output

Change-Id: Iab63a9cd398e975f5099de87022fceaf9999575a
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

jonathanhaley commented:
Hi @powerjg I think this is ready for a proper review. I have implemented your initial requested changes

giactra commented:
@powerjg can we get a revie when you get some time?
